htlcswitch/switch.go

```diff
@@ -889,64 +903,80 @@ func (s *Switch) handleLocalResponse(pkt *htlcPacket) {
 	// has been restarted since sending the payment.
 	payment := s.findPayment(pkt.incomingHTLCID)
 
-	var (
-		preimage   [32]byte
-		paymentErr error
-	)
+	// The error reason will be unencypted in case this a local
+	// failure or a converted error.
+	unencrypted := pkt.localFailure || pkt.convertedError
+	n := &networkResult{
+		msg:          pkt.htlc,
+		unencrypted:  unencrypted,
+		isResolution: pkt.isResolution,
+	}
 
-	switch htlc := pkt.htlc.(type) {
+	// Deliver the payment error and preimage to the application, if it is
+	// waiting for a response.
+	if payment != nil {
+		payment.resultChan <- n
+		s.removePendingPayment(pkt.incomingHTLCID)
```

Comment:
seems like there's a possible race where:
1. `SendHTLC` is called, and the pending payment is added to the map
2. The HTLC is settled/failed (most likely, failed locally) and the pending payment is removed from the map
3. `GetPaymentResult` is called but the pending payment is no longer found

Perhaps we can just opt to not remove pending payments from the map? Would think the flow would be more like:
1. If pending payment exists, wait for ntfn to be delivered via channel
2. If no pending payment exists, retrieve from database or cached in-memory response

Reply:
Yeah this doesn't yet resolve races like this or gaps in persistence. Those will be taken care of in  #2762 (the pending payments map goes away all together).

Reply:
gotcha, okay looking through #2762 this race does seem to be eliminated

Reply:
It's a good point, it does introduce a race. This is fixed in #2762, but it is a bummer that this PR introduces the race. Temp fixed it by not removing the pending payment at all: https://github.com/lightningnetwork/lnd/pull/3063/commits/1f9b4fb260979920cb62eda3e9940384becdc1bd

Reply:
@halseth nice, i think that is okay. if for some reason this happens twice then it block everything since the channel is only buffered with 1 item. could also just temporarily buffer them at 100 to be safe, but i think the deduping by the circuit map should prevent this.